Commit the draft (#1)
LLFourn/witness-asymmetric-channel: README.md

_This is draft email to the lightning-dev mailing list. If all goes well maybe it can also be a place iterate on the design._

# Background

As presently specified, the two parties in a lightning channel are assigned different commitment transactions.
This is logically necessary for the protocol to identify which party broadcasted the commitment transaction and potentially punish them if the other party provides provides proof it has been revoked (i.e. knows the revocation key).

Wouldn't it be nice if we could identify the broadcasting party without assigning them different transactions?

> Riard first considered this problem in [8] and proposed that you could identify the party using _witness asymmetry_ i.e. both parties broadcast the same transaction but have different witnesses. 

Reviewer: And combining penalty support with eltoo-kind of channel upgrades. IIRC scheme was "update-and-punish", conceptually separating both operations (but still limited to 2 parties)

Author: I expanded slightly to mention motivation for your idea: https://github.com/LLFourn/witness-asymmetric-channel/pull/1/commits/65b5a02f0957b93ed14d97a66e598a4eab81a3a9

The adaptor signatures force the party broadcasting a commitment transaction to reveal a "publishing secret" to the other party.
The honest party can then use this publishing secret along with knowledge of the usual revocation secret to punish the malicious party for broadcasting an old commitment transaction.

In the paper this idea is combined with "punish-then-split" mechanism similar to the original eltoo[8] proposal and therefore it unfortunately inherits the issue with staggering time-locks.
BOLT 3 [4] notes this problem as design motivation:

> The reason for the separate transaction stage for HTLC outputs is so that HTLCs can timeout or be fulfilled even though they are within the to_self_delay delay. Otherwise, the required minimum timeout on HTLCs is lengthened by this delay, causing longer timeouts for HTLCs traversing the network.

More in depth commentary by Towns can be found in [2] and they propose a solution to the eltoo proposal in [3].

In short, the problem is that it creates a risk for the party that needs to fulfill an HTLC with the secret in time.
The only known way of accounting for this risk is to increase the difference between the time-locks on each hop.

> There could be situations where this trade off makes sense but it seems like it undesirable for a general purpose payment channel network.

Reviewer: "as it would make any HTLC-targeting attack easier" like flood & loot, pinning, time-dilation.

Recent trend has been to increase them, https://github.com/lightningnetwork/lightning-rfc/pull/785

Author: I'm not sure I can justify that claim. If you account for the relative time-lock by extending the absolute time-lock doesn't that address this (I mean in so far as it's not worse than the current state of a affairs)?

I have a feeling you are right in practice but I don't see the mechanics of it. The analysis of these attacks is certainly more difficult with the punish-then-settle model.
It might be time to do some math and stop using feelings though!

# Proposal

I propose a channel protocol which steals the main idea from Aumayr et al. while retaining the time-lock precedence of BOLT 3 transactions.
That is, absolute time-locks are settled before relative time-locks to avoid having to account for relative time-locks when calculating absolute time-locks.

The guiding principal is to keep commitment transactions symmetric but to use the asymmetric knowledge of the parties to simulate the way lightning works right now.
The main benefits of this seem to be:

> - It is more elegant as there are half the number of possible transactions. I expect this will follow through to reduced implementation complexity and maybe make it easier to explain as well.

Reviewer: Yes it should simplify onchain transaction parser and thus increase their reliablity.

- Every output can just be a 2-of-2 multi-sig (which could be a single key with MuSig or OP_CHECKMULTISIG with ECDSA etc). This makes the transactions a little smaller even when using 2-of-2 with ECDSA on balance (I think). The space saving comes from avoiding explicit revocation keys in the commitment transaction outputs.

> - The design works today and easily transitions into a post-taproot world although there is still a bit to explore there [5].

Reviewer 2: Not sure if this is correct though :) 
```suggestion
- The design works today and easily transitions into a post-taproot world although there is still a bit exploration needed [5].
```

Author: I think the informal way the original is worded is ok for a mailing list post.

## Commitment transaction outputs

> The way the fund transaction is constructed and negotiated remains unchanged.

Reviewer 2: Maybe it's good to refer back to the paper or shortly explain the concept of the _funding transaction_ here?

Author: Yes done.

For each commitment transaction parties exchange two curve points each:

- A _publication_ point (denoted P) whose secret is revealed if that party broadcasts the commitment transaction.
- A _revocation_ point (denoted R) whose secret is explicitly revealed to the other party when revoking this commitment transaction.

I'll use the notation (Pa, Ra) and (Pb, Rb) to refer to the publication and revocation points of Alice and Bob respectively and 2-of-2(X,Y) to refer to a multi-signature output that supports adaptor signatures and requires the owners of X and Y to authorise it (e.g. `OP_CHECKMULTISIG` with ECDSA [6], or MuSig with Schnorr etc [7]).

The scriptPubkey of all outputs on a commitment transaction is 2-of-2(Pa + Ra, Pb + Rb). 
For a slight increase in fungibility you can imagine that we pseudorandomly randomize each output so that they don't all look identical to blockchain observers.

When signing a commitment transaction, the parties instead exchange adaptor signatures such that the party who publishes the commitment transaction must reveal their publication secret.

> This means that if a party broadcasts a commitment transaction for which they have already revealed their revocation secret the other party will know both secrets and therefore both secret keys in 2-of-2(Pa + Ra, Pb + Rb).

Reviewer 2: I'm familiar with the paper and know what you mean but maybe it's worth mentioning that we assume that only a malicious party would publish an old state and hence broadcast an old commitment transaction?  
```suggestion
This means that if a malicious party broadcasts a commitment transaction for which they have already revealed their revocation secret the other party will know both secrets and therefore both secret keys in 2-of-2(Pa + Ra, Pb + Rb).
```

Author: Yes. Done.

They can use this to unilaterally spend from all the outputs on the commitment transaction to punish the malicious party.

> Note there is no problem with both parties broadcasting the same commitment transaction at the same time with their different signatures.

Reviewer 2: Isn't the problem that only one of the transactions can be included? 🙃 This means that both parties need to monitor their node's mempool to learn about the other party's transaction and publication secret.

Author: > Isn't the problem that only one of the transactions can be included?

The point of this sentence is to say this isn't a problem.

>  This means that both parties need to monitor their node's mempool to learn about the other party's transaction and publication secret.

They don't need to but if they do they can settle the state _more_ efficiently than they otherwise would be able to .

Both parties seeing the other's commitment transaction over the Bitcoin network actually acts as a kind of out of band optimization because it means they both learn each other's publication secrets and can therefore settle all the outputs without having to wait for relative time-locks (regardless of which signature makes it into the blockchain).

## Balance outputs

For a balance output going to Alice the output is 2-of-2(Pb, Ra) and the converse for Bob's balance output i.e. 2-of-2(Pa,Rb).

> Each balance output has a pre-signed relative time-locked claim transaction spending it to an address owned by Alice.

Reviewer 2: Do you really mean that `each balance output [...]` goes to `[...] an address owned by Alice`? 

Shouldn't it be either: 

```suggestion
Each balance output has a pre-signed relative time-locked claim transaction spending it to an address of the deserving party.
```

or: 

```suggestion
A balance output for Alice has a pre-signed relative time-locked claim transaction spending it to an address owned by Alice (and converse for Bob).
```

Author: Yes fixed.
